crates/sui-core/src/authority_server.rs

```diff
@@ -540,6 +542,198 @@ impl ValidatorService {
+            cont.push(request_idx);
+        }
+
+        // Second pass
+        let mut verified_certs = Vec::new();
+        let mut certs_idx_to_submit = Vec::new();
+        // 3) All certificates are sent to consensus (at least by some authorities)
+        // For shared objects this will wait until either timeout or we have heard back from consensus.
+        // For owned objects this will return without waiting for certificate to be sequenced
+        // First do quick dirty non-async check
+        for _request_idx in &cont {
+            let certificate = cert_verify_futures.remove(0).await?;
+            if !epoch_store.is_tx_cert_consensus_message_processed(&certificate)? {
+                certs_idx_to_submit.push(verified_certs.len());
+            }
+            verified_certs.push(certificate);
+        }
+        assert!(cert_verify_futures.is_empty());
+
+        {
+            let reconfiguration_lock = epoch_store.get_reconfig_state_read_lock_guard();
+            if !reconfiguration_lock.should_accept_user_certs() {
+                metrics
+                    .num_rejected_cert_in_epoch_boundary
+                    .inc_by(certs_idx_to_submit.len() as u64);
+                return Err(SuiError::ValidatorHaltedAtEpochEnd.into());
+            }
+            if certs_idx_to_submit.len() > 0 {
+                let _metrics_guard = metrics.consensus_latency.start_timer();
+                let transactions = certs_idx_to_submit
+                    .into_iter()
+                    .map(|cert_idx| {
+                        ConsensusTransaction::new_certificate_message(
+                            &state.name,
+                            verified_certs[cert_idx].clone().into(),
+                        )
+                    })
+                    .collect();
+                consensus_adapter.submit_batch(
+                    transactions,
+                    Some(&reconfiguration_lock),
+                    &epoch_store,
+                )?;
+            }
+            drop(reconfiguration_lock);
+        }
+
+        // 4) Execute the certificate if it contains only owned object transactions, or wait for
+        // the execution results if it contains shared objects.
+        for (cert_idx, request_idx) in cont.into_iter().enumerate() {
+            let effects = state
+                .execute_certificate(&verified_certs[cert_idx], &epoch_store)
+                .await?;
+            let events = if let Some(event_digest) = effects.events_digest() {
+                state.get_transaction_events(event_digest)?
+            } else {
+                TransactionEvents::default()
+            };
+            responses[request_idx] = Some(HandleCertificateResponseV2 {
+                signed_effects: effects.into_inner(),
+                events,
+                fastpath_input_objects: vec![], // unused field
+            });
+        }
+
+        for response in &responses {
+            assert!(response.is_some());
```

Comment:
If you create the vector with with_capacity instead of resize, then you shouldn't need to use None and the assert here.